didcomm_messaging/quickstart.py

```diff
@@ -0,0 +1,393 @@
+    # DIDCommMessaging handles the orchestration of each individual service,
+    # ensuring that messages get packed and delivered via a simple and straight
+    # forward interface.
+    DMP = DIDCommMessaging(
+        crypto=crypto,
+        secrets=secrets,
+        resolver=resolver,
+        packaging=packer,
+        routing=router,
+    )
+
+    return DMP
+
+
+async def send_http_message(
+    dmp: DIDCommMessaging, my_did: DID, message: Message, target: DID
+):
+    """Send a message via HTTP."""
+
+    # Get the message as a dictionary
+    message_wrapper = message
+    message = message_wrapper.as_dict()
+
+    # Encrypt/pack the message to our target
+    packy = await dmp.pack(
+        message=message,
+        to=target,
+        frm=my_did,
+    )
+    packed = packy.message
+
+    # Get the first http endpoint from the last DID in the DID chain
+    endpoint = packy.get_endpoint("http")
+
+    async with aiohttp.ClientSession() as session:
+        LOG.info("posting message type %s to %s" % (message_wrapper.type, endpoint))
+
+        async with session.post(endpoint, data=packed) as resp:
+            LOG.debug("posted message: %s" % (message))
+            LOG.debug("message ID: %s" % (message_wrapper.id))
+            packed = await resp.text()
+
+            # If the HTTP enpoint responded with a message, decode it
+            if len(packed) > 0:
+                unpacked = await dmp.packaging.unpack(packed)
+                msg = unpacked[0].decode()
+                LOG.debug("Raw message from remote %s" % msg)
+                return Message.from_json(msg)
+    return
+
+
+async def setup_relay(
+    dmp: DIDCommMessaging, my_did: DID, relay_did: DID, keys: Sequence[Key]
+) -> Union[DID, None]:
+    """Negotiate services with an outbound relay.
+
+    Returns a DID upon successful negotiation.
+    """
+
+    # Request mediation from the outbound relay
+    message = Message(
+        type="https://didcomm.org/coordinate-mediation/3.0/mediate-request",
+        id=str(uuid.uuid4()),
+        body={},
+        frm=my_did,
+        to=[relay_did],
+    )
+    message = await send_http_message(dmp, my_did, message, target=relay_did)
+
+    # Verify that mediation-access has been granted
+    if message.type == "https://didcomm.org/coordinate-mediation/3.0/mediate-deny":
+        return
+    if message.type != "https://didcomm.org/coordinate-mediation/3.0/mediate-grant":
+        # We shouldn't run into this case, but it's possible
+        raise Exception("Unknown response type received: %s" % message.type)
+
+    # Create a new DID with an updated service endpoint, pointing to our relay
+    relay_did = message.body["routing_did"][0]
+    new_did = generate(
+        keys,
+        [
+            {
+                "type": "DIDCommMessaging",
+                "serviceEndpoint": {
+                    "uri": relay_did,
+                    "accept": ["didcomm/v2"],
+                },
+            }
+        ],
```

Comment:
Most of this is static; should we just have a parameter that takes the `relay_did` and makes the rest of that in the generate method?

Reply:
If we have the `generate` method in did:peer:N allow either a DID or a service structure, I could agree. However, showing the structure here does also provide the benefit of teaching people that "this is what a basic service block looks like when you generate a DID"

Reply:
I think we get that "this is what it looks like" by having the `generate` method nearby in the quickstart module (which it is).

Reply:
Since the generate method is defined elsewhere, I have created two issues regarding this change:
- https://github.com/decentralized-identity/did-peer-2/issues/4
- https://github.com/decentralized-identity/did-peer-4/issues/8

The solution should be handled in their respective projects.